File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.java
class Solution {

public int smallestDivisor(int[] nums, int threshold) {

    int s = 1;
    int e = Integer.MIN_VALUE;

    for (int num : nums) {
        e = Math.max(e, num);
    }

    int ans = 0;

    if (nums.length == threshold) return e;

    while (s <= e) {

        int m = s + (e - s) / 2;
        int total = 0;

        for (int num : nums) {
            total += (num + m - 1) / m;   // ceiling division
        }

        if (total <= threshold) {
            ans = m;
            e = m - 1;
        } else {
            s = m + 1;
        }
    }

    return ans;
}

}

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int smallestDivisor(vector<int>& nums, int threshold) {
        int start = 1 , end = *max_element(nums.begin() , nums.end());
        int ans = -1;
        while(start <= end){
            int mid = start + ( end - start )/ 2;
            int total = 0;
            for(int num : nums){
                total += ceil(double(num) / mid);
            }
            if(total <= threshold){
                ans = mid;
                end = mid - 1;
            }else start = mid + 1 ;
        }
        return ans;
    }
};